### ISSUE TITLE
pg_probackup process killed during PAGE backup

### ISSUE THREAD

**Issue by beremour** [postgrespro/pg_probackup#252]
I have repository:
```
bash-4.2$ pg_probackup-12 show     -B /backup/pgsql
BACKUP INSTANCE 'main'

========================================================================================================================================
 Instance  Version  ID      Recovery Time           Mode  WAL Mode  TLI    Time   Data    WAL  Zratio  Start LSN    Stop LSN     Status
========================================================================================================================================
 main      12       QFW4NN  2020-08-30 22:58:29+03  FULL  STREAM    2/0  1h:32m  496GB   50GB    1.00  89/E8870760  8D/4870FE80  OK
 main      12       QFUZQN  2020-08-30 07:13:00+03  PAGE  STREAM    2/2     23s  150MB  192MB    1.00  2/B000110    2/1659BB70   OK
 main      12       QFUZOL  2020-08-30 07:12:27+03  FULL  STREAM    2/0     12s  182MB   16MB    1.00  2/B000110    2/B000250    OK
-bash-4.2$

```
When I try to take PAGE-backup from STANDBY instance, process get killed abnormaly:
```
-bash-4.2$ pg_probackup-12 backup   -B /backup/pgsql -U backup --instance main -b PAGE --stream --temp-slot --merge-expired --delete-expired --delete-wal -j 16 --progress --log-directory=/vol/pgsql --log-filename=probackup-page-%y%m%d-%H%M%S.log --log-level-file=verbose
INFO: Backup start, pg_probackup version: 2.4.2, instance: main, backup ID: QFXDZ8, backup mode: PAGE, wal mode: STREAM, remote: false, compress-algorithm: none, compress-level: 1
INFO: Backup QFXDZ8 is going to be taken from standby
INFO: Wait for WAL segment /backup/pgsql/wal/main/000000020000010D0000002E to be archived
INFO: Parent backup: QFW4NN
INFO: PGDATA size: 936GB
INFO: Extracting pagemap of changed blocks
Killed
-bash-4.2$
```
log file is: [probackup-page-200831-141620.log](https://github.com/postgrespro/pg_probackup/files/5150070/probackup-page-200831-141620.log)

WAL directory in backup repository has large enouph count of files:
```
[root@pgbase02 ~]# ls -l /backup/pgsql/wal/main | wc -l
68389
[root@pgbase02 ~]#
```
Server has 64GB of memory:
```
[root@pgbase02 ~]# free -m
              total        used        free      shared  buff/cache   available
Mem:          64429        2884       44867       16137       16676       44632
Swap:           763         760           3
[root@pgbase02 ~]#
```
Can you please help with it ?

**Comment by gsmol** [comment]
Hello!
Hm, looks like an oom-killer work.
Can you please provide the output for `dmesg -T` command ?

**Comment by beremour** [comment]
Look please: [dmesg-T.txt](https://github.com/postgrespro/pg_probackup/files/5150128/dmesg-T.txt)

**Comment by gsmol** [comment]
Well that is serious, looks like we are leaking memory: `anon-rss:45740328kB`.
Can you please run again backup with `--log-level-file=VERBOSE --log-filename=pbk-leak.log` options and attach file ` /backup/pgsql/log/pbk-leak.log`?

**Comment by beremour** [comment]
Log file in the 1st post already created with these options:

pg_probackup-12 backup   -B /backup/pgsql -U backup --instance main -b PAGE --stream --temp-slot --merge-expired --delete-expired --delete-wal -j 16 --progress --log-directory=/vol/pgsql --log-filename=probackup-page-%y%m%d-%H%M%S.log --log-level-file=verbose

Shall I do it again ?

**Comment by gsmol** [comment]
> Log file in the 1st post already created with these options
>
Ah, thank you, I`ve missed that.
Looks like WAL parsing routine is leaking memory, I will look into it.

**Comment by gsmol** [comment]
Can you please provide the output of the following command:
```
pg_probackup-12 show -B /backup/pgsql --archive
```
?

**Comment by gsmol** [comment]
Is it reproducible?

**Comment by beremour** [comment]
> Can you please provide the output of the following command:
> 
> ```
> pg_probackup-12 show -B /backup/pgsql --archive
> ```
> 
> ?

```
-bash-4.2$ pg_probackup-12 show -B /backup/pgsql --archive

ARCHIVE INSTANCE 'main'
=================================================================================================================================
 TLI  Parent TLI  Switchpoint  Min Segno                 Max Segno                 N segments  Size    Zratio  N backups  Status
=================================================================================================================================
 2    0           0/0          00000002000000020000000B  000000020000010D00000099  68495       1070GB  1.00    8          OK
-bash-4.2$
```

**Comment by gsmol** [comment]
Can you please provide the `/backup/pgsql/backups/main/QFW4NN/backup_content.control` file?

**Comment by beremour** [comment]
> Is it reproducible?

Error in 1st post occurs after:
- large data insertion to the master host
- full backup at standby host

PAGE backup has huge "Extracting pagemap of changed blocks" phase.
 
Since 1st post I've made:
- Configure huge pages and reboot server. Now postgres instance is in huge pages area.
- Several page backups made

Now PAGE backup has shot "Extracting pagemap of changed blocks" phase

**Comment by beremour** [comment]
Do you interested in postgres settings ?

**Comment by beremour** [comment]
> /backup/pgsql/backups/main/QFW4NN/backup_content.control
[backup_content.control.zip](https://github.com/postgrespro/pg_probackup/files/5157648/backup_content.control.zip)

**Comment by beremour** [comment]
Current repository:
```
-bash-4.2$ pg_probackup-12 show -B /backup/pgsql

BACKUP INSTANCE 'main'
=============================================================================================================================================
 Instance  Version  ID      Recovery Time           Mode  WAL Mode  TLI    Time    Data     WAL  Zratio  Start LSN     Stop LSN      Status
=============================================================================================================================================
 main      12       QFZMB1  2020-09-01 19:11:28+03  PAGE  STREAM    2/2  3m:22s  1497MB  1456MB    1.00  10D/AA2A2A00  10E/49B2050   OK
 main      12       QFZLYT  2020-09-01 16:15:12+03  PAGE  STREAM    2/2  2m:16s  1497MB  1328MB    1.00  10D/49036908  10D/9AEF0E40  OK
 main      12       QFZE67  ----                    PAGE  STREAM    2/2  3m:13s   124kB       0    1.00  10D/49036908  0/0           ERROR
 main      12       QFZCZA  2020-09-01 15:52:18+03  PAGE  STREAM    2/2  3m:12s   134kB    32MB    1.00  10D/2F001A48  10D/30000060  DONE
 main      12       QFY2ZK  2020-09-01 00:22:09+03  FULL  STREAM    2/0   1h:8m   937GB    16MB    1.00  10D/2F001718  10D/2F0017C8  OK
 main      12       QFXSQI  2020-08-31 20:55:13+03  FULL  STREAM    2/0  1h:22m   937GB    16MB    1.00  10D/2F001718  10D/2F0017C8  OK
 main      12       QFXDZ8  ----                    PAGE  STREAM    2/2   1d:5h       0       0    1.00  10D/2E000060  0/0           RUNNING
 main      12       QFW4NN  2020-08-30 22:58:29+03  FULL  STREAM    2/0  1h:32m   496GB    50GB    1.00  89/E8870760   8D/4870FE80   OK
 main      12       QFUZQN  2020-08-30 07:13:00+03  PAGE  STREAM    2/2     23s   150MB   192MB    1.00  2/B000110     2/1659BB70    OK
 main      12       QFUZOL  2020-08-30 07:12:27+03  FULL  STREAM    2/0     12s   182MB    16MB    1.00  2/B000110     2/B000250     OK
-bash-4.2$ pg_probackup-12 show -B /backup/pgsql --archive

ARCHIVE INSTANCE 'main'
=================================================================================================================================
 TLI  Parent TLI  Switchpoint  Min Segno                 Max Segno                 N segments  Size    Zratio  N backups  Status
=================================================================================================================================
 2    0           0/0          00000002000000020000000B  000000020000010E00000003  68601       1072GB  1.00    10         OK
-bash-4.2$
```

**Comment by gsmol** [comment]
> Do you interested in postgres settings ?
>
At this moment - no, but thank you.

Can you also provide `/backup/pgsql/backups/main/QFW4NN/backup_content.control` file ?

**Comment by beremour** [comment]
> Can you also provide `/backup/pgsql/backups/main/QFW4NN/backup_content.control` file ?

File in attached in 2 posts earlier isn't OK ?

**Comment by gsmol** [comment]
> Now PAGE backup has shot "Extracting pagemap of changed blocks" phase.
>
It was forced to process 500GB of WAL, for every changed file there is a bitmap in memory, which should be 16kB at most, so memory consumption should be at 100-200MB tops.
And it processed just a fraction of total WAL (~50GB) before been killed by oom for nomming ~45GB of memory.

Doesn`t look like a memory leak.

**Comment by gsmol** [comment]
> File in attached in 2 posts earlier isn't OK ?
>
I`m sorry, I was mean to ask for `/backup/pgsql/backups/main/QFXSQI/backup_content.control`.

**Comment by beremour** [comment]
here it is: [backup_content.control.zip](https://github.com/postgrespro/pg_probackup/files/5158125/backup_content.control.zip)

**Comment by gsmol** [comment]
Also, is the backup catalog by any change located on nfs mount ? If so, can you provide the mount options?

**Comment by beremour** [comment]
They are just defauts:
```
-bash-4.2$ mount | grep -i backup
/etc/auto.direct on /backup type autofs (rw,relatime,fd=5,pgrp=8635,timeout=300,minproto=5,maxproto=5,direct,pipe_ino=39478)
pgbeckup:/vol on /backup type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.3.172,local_lock=none,addr=192.168.3.173)
-bash-4.2$
```

**Comment by gsmol** [comment]
Thank you!

**Comment by beremour** [comment]
More info from standby server, once again huge "Extracting pagemap" phase
```
-bash-4.2$ /usr/bin/time pg_probackup-12 backup   -B /backup/pgsql -U backup --instance main -b PAGE --stream --temp-slot --merge-expired --delete-expired --delete-wal -j 16 --log-directory=/vol/pgsql --log-filename=probackup-page-%y%m%d-%H%M%S.log --log-level-file=verbose --no-validate --skip-block-validation
INFO: Backup start, pg_probackup version: 2.4.2, instance: main, backup ID: QG04N4, backup mode: PAGE, wal mode: STREAM, remote: false, compress-algorithm: none, compress-level: 1
INFO: Backup QG04N4 is going to be taken from standby
INFO: Wait for WAL segment /backup/pgsql/wal/main/0000000200000122000000E4 to be archived
WARNING: Backup QG045S has status: RUNNING. Cannot be a parent.
WARNING: Backup QG03ZH has status: ERROR. Cannot be a parent.
WARNING: Backup QG02Y4 has status: ERROR. Cannot be a parent.
INFO: Parent backup: QFZWWU
INFO: PGDATA size: 949GB
INFO: Extracting pagemap of changed blocks
out of memory
out of memory
out of memory
out of memory
out of memory
426.97user 372.82system 5:03.47elapsed 263%CPU (0avgtext+0avgdata 24511524maxresident)k
122982034inputs+101801outputs (1major+6128022minor)pagefaults 0swaps
```
```
-bash-4.2$ free -m
              total        used        free      shared  buff/cache   available
Mem:          64429       17737       28127           8       18564       46129
Swap:           763           4         759
-bash-4.2$
```
here is verbose logfile: [probackup-page-200902-014728.log](https://github.com/postgrespro/pg_probackup/files/5159177/probackup-page-200902-014728.log)
```
-bash-4.2$ pg_probackup-12 show -B /backup/pgsql

BACKUP INSTANCE 'main'
=============================================================================================================================================
 Instance  Version  ID      Recovery Time           Mode  WAL Mode  TLI     Time   Data     WAL  Zratio  Start LSN     Stop LSN      Status
=============================================================================================================================================
 main      12       QG04N4  ----                    PAGE  STREAM    2/2  11m:41s      0       0    1.00  122/E4000060  0/0           RUNNING
 main      12       QG045S  ----                    PAGE  STREAM    2/2   22m:5s      0       0    1.00  122/E25EB200  0/0           RUNNING
 main      12       QG03ZH  ----                    PAGE  STREAM    2/2      37s      0       0    1.00  122/E25EB200  0/0           ERROR
 main      12       QG02Y4  ----                    PAGE  STREAM    2/2    5m:1s      0       0    1.00  122/E25EB118  0/0           ERROR
 main      12       QFZWWU  2020-09-01 22:29:03+03  PAGE  STREAM    2/2  10m:25s   17GB    11GB    1.00  112/80155750  115/573CB0F0  OK
 main      12       QFZVMI  2020-09-01 22:34:40+03  PAGE  STREAM    2/2   6m:12s   14GB  9152MB    1.00  10F/E76AAB58  112/209788C8  OK
 main      12       QFZMB1  2020-09-01 19:11:28+03  FULL  STREAM    2/0      46s  937GB  1456MB    1.00  10D/AA2A2A00  10E/49B2050   OK
 main      12       QFZE67  ----                    PAGE  STREAM    2/2   3m:13s  124kB       0    1.00  10D/49036908  0/0           ERROR
 main      12       QFXSQI  2020-08-31 20:55:13+03  FULL  STREAM    2/0   1h:22m  937GB    16MB    1.00  10D/2F001718  10D/2F0017C8  OK
 main      12       QFXDZ8  ----                    PAGE  STREAM    2/2   1d:11h      0       0    1.00  10D/2E000060  0/0           RUNNING
 main      12       QFW4NN  2020-08-30 22:58:29+03  FULL  STREAM    2/0   1h:32m  496GB    50GB    1.00  89/E8870760   8D/4870FE80   OK
 main      12       QFUZQN  2020-08-30 07:13:00+03  PAGE  STREAM    2/2      23s  150MB   192MB    1.00  2/B000110     2/1659BB70    OK
 main      12       QFUZOL  2020-08-30 07:12:27+03  FULL  STREAM    2/0      12s  182MB    16MB    1.00  2/B000110     2/B000250     OK
-bash-4.2$
```
I will freeze DBs now

**Comment by gsmol** [comment]
Would it be too much to ask to setup automatic core dump generation?

**Comment by gsmol** [comment]
Our current suspect is fragmentation in bitmap reallocation code, which is coming from PostgreSQL code.
We are going to rewrite it and, if you are willing, provide you with binary to try.

**Comment by beremour** [comment]
> Would it be too much to ask to setup automatic core dump generation?

How to do it ?

**Comment by beremour** [comment]
> Our current suspect is fragmentation caused by bitmap reallocation code, which is coming from PostgreSQL code.
> We are going to rewrite it and, if you are willing, provide you with binary to try.

I wll surely try new binary

**Comment by gsmol** [comment]
> How to do it ?
>
That will be out plan B.

> I wll surely try new binary
>
What operation system do you use?

**Comment by beremour** [comment]
```
# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.6 (Maipo)
# uname -a
Linux pgbase02 3.10.0-957.el7.x86_64 #1 SMP Thu Oct 4 20:48:51 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
#
```

**Comment by gsmol** [comment]
Packages with new binaries:
https://oc.postgrespro.ru/index.php/s/Tyk5FjazhDyD5Bl

Memory for bitmap now allocated using different algorithm.

**Comment by beremour** [comment]
Sorry, but result is the same:
```
-bash-4.2$ /usr/bin/time pg_probackup-12 backup   -B /backup/pgsql -U backup --instance main -b PAGE --stream --temp-slot --merge-expired --delete-expired --delete-wal -j 16 --log-directory=/vol/pgsql --log-filename=probackup-page-%y%m%d-%H%M%S.log --log-level-file=verbose --no-validate --skip-block-validation
INFO: Backup start, pg_probackup version: 2.4.3, instance: main, backup ID: QG8ZTN, backup mode: PAGE, wal mode: STREAM, remote: false, compress-algorithm: none, compress-level: 1
INFO: Backup QG8ZTN is going to be taken from standby
WARNING: Backup QG3HF8 has status: RUNNING. Cannot be a parent.
INFO: Parent backup: QFZWWU
INFO: PGDATA size: 949GB
INFO: Extracting pagemap of changed blocks
out of memory
out of memory
out of memory
out of memory
out of memory
380.00user 240.12system 4:29.75elapsed 229%CPU (0avgtext+0avgdata 24471492maxresident)k
122980120inputs+182439outputs (0major+6117981minor)pagefaults 0swaps
```
```
-bash-4.2$ free -m
              total        used        free      shared  buff/cache   available
Mem:          64429       17676       27503          48       19249       46122
Swap:           763           4         759
-bash-4.2$
```
```
-bash-4.2$ ls -l /usr/bin/pg_probackup-12
-rwxr-xr-x. 1 root root 410760 Sep  6 14:55 /usr/bin/pg_probackup-12
-bash-4.2$ /usr/bin/time pg_probackup-12 --version
pg_probackup-12 2.4.3 (PostgreSQL 12.4)
```
Verboze log is:  [probackup-page-200906-204147.log](https://github.com/postgrespro/pg_probackup/files/5180050/probackup-page-200906-204147.log)

**Comment by gsmol** [comment]
I`ve added some additional debug messages to confirm/refute our current version. Can you run PAGE backup again with new binary?
Packages with new binaries:
https://oc.postgrespro.ru/index.php/s/Tyk5FjazhDyD5Bl

**Comment by gsmol** [comment]
Also can you run concurently `atop` program with `-a 1` and `-w` options, so we can see the progress of memory consumption more clearly?

**Comment by gsmol** [comment]
Good news, I`ve managed to reproduce the issue, so requests from previous posts can be ignored.

**Comment by gsmol** [comment]
Packages with new binaries, should fix the issue:
https://oc.postgrespro.ru/index.php/s/Tyk5FjazhDyD5Bl

**Comment by beremour** [comment]
Wow! Seems it was a great hunting :) You binary really solves this terrible issue
Verbose log of successfull page backup: [probackup-page-200910-082403.log](https://github.com/postgrespro/pg_probackup/files/5199467/probackup-page-200910-082403.log)

Should I just ignore following "WARNING : Could not read WAL record" ?
```
2020-09-10 08:31:41 MSK [24864]: LOG: Thread [0]: Opening WAL segment "/backup/pgsql/backups/main/QGFGC3/database/pg_wal/0000000200000126000000FE"
2020-09-10 08:31:41 MSK [24864]: WARNING: Could not read WAL record at 126/FE7604C0: invalid record length at 126/FE7604C0: wanted 24, got 0
2020-09-10 08:31:41 MSK [24864]: VERBOSE: Thread [0]: Need to switch to the next WAL segment, page LSN 126/FE000000, record being read LSN 0/0
2020-09-10 08:31:41 MSK [24864]: LOG: Thread [0]: Opening WAL segment "/backup/pgsql/backups/main/QGFGC3/database/pg_wal/0000000200000126000000FE"
2020-09-10 08:31:41 MSK [24864]: LOG: Record 126/FE760488 has endpoint 126/FE7604C0 which is equal or greater than requested LSN 126/FE7604C0
2020-09-10 08:31:41 MSK [24864]: LOG: Found prior LSN: 126/FE760488
2020-09-10 08:31:43 MSK [24864]: LOG: finished streaming WAL at 126/FE7604C0 (timeline 2)
2020-09-10 08:31:43 MSK [24864]: LOG: Getting the Recovery Time from WAL
```

**Comment by gsmol** [comment]
> Wow! Seems it was a great hunting :)
>
Well, actually it was quite confusing, for some reason fsanitizer repeatedly failed to detect leak when you first reported this issue, which threw us to a wrong conclusion about problem been in bitmap code somewhere. But after we found the real issue by using the method of `looking really hard into the code`  fsanitizer started to detect leak =)

> Should I just ignore following "WARNING : Could not read WAL record" ?
>
That is ok for a backup from replica. For some reason PostgreSQL thinks it is ok to return LSN of non-existing WAL record as STOP LSN, so when pg_probackup tries to ensure that all WAL, required for successful recovery is available, it fails for find it. So we have to dance around it, first waiting for this record to arrive (that is where all this `Could not read WAL record at 126/FE7604C0` messages come from), then by using previous WAL record as STOP LSN.

**Comment by gsmol** [comment]
Я бы предложил увеличить этот параметр до, скажем, 10 минут.

**Comment by beremour** [comment]
I hope It helps make project more stable!
Can you please explain how to deal with "Wait for WAL segment /vol/pgsql/wal/main/000000020000012700000074 to be archived" ? To force WAL-switching on Master is the only way ?

**Comment by beremour** [comment]
```
INFO: Backup QGFVQZ is going to be taken from standby
INFO: Wait for WAL segment /vol/pgsql/wal/main/000000020000012700000074 to be archived
ERROR: WAL segment 000000020000012700000074 could not be archived in 300 seconds
WARNING: backup in progress, stop backup
WARNING: Backup QGFVQZ is running, setting its status to ERROR
```

**Comment by gsmol** [comment]
> I hope It helps make project more stable!
>
We really appreciate your feedback and efforts, thanks!

> Can you please explain how to deal with "Wait for WAL segment /vol/pgsql/wal/main/000000020000012700000074 to be archived" ? To force WAL-switching on Master is the only way ?
>
Is this issue happens during the period of low WAL traffic coming from master?

**Comment by beremour** [comment]
I run it just now to be sure and Yes WAL traffic is low.

**Comment by gsmol** [comment]
> I run it just now to be sure and Yes WAL traffic is low.
>
The most straightforward way to fix this is to set `archive_timeout` PostgreSQL option to `5min` and pg_probackup `--archive-timeout` option to `10min`. PostgreSQL in case of `>5min` inactivity will forcibly switch WAL segment into the archive, where it can be used by pg_probackup.
The more elegant (but also more demanding) solution is to use `pg_receivewal` utility instead of `archive_command` to populate the WAL archive.

**Comment by beremour** [comment]
Currently I am using ```archive_command``` as ``` pg_probackup-12 archive-push```, but thank you, I will check possibility to run ``` pg_receivewal``` on backup server. But "timeout"-way also seems acceptable.

**Comment by gsmol** [comment]
`archive_timeout` PostgreSQL option should be enabled on master, of course.

**Comment by gsmol** [comment]
Fix will shipped with 2.4.4 release.
Is it ok to close this issue?

**Comment by beremour** [comment]
Thank you very much!
